File: Data-Structure/Java/LeetCode/Queue and Stack/286. Walls and Gates.java
//286. Walls and Gates
/*
You are given an m x n grid rooms initialized with these three possible values.

-1 A wall or an obstacle.
0 A gate.
INF Infinity means an empty room. We use the value 2^31 - 1 = 2147483647 to represent INF as you may assume that the distance to a gate is less than 2147483647.
Fill each empty room with the distance to its nearest gate. If it is impossible to reach a gate, it should be filled with INF. 

Input: rooms = [[2147483647,-1,0,2147483647],[2147483647,2147483647,2147483647,-1],[2147483647,-1,2147483647,-1],[0,-1,2147483647,2147483647]]
Output: [[3,-1,0,1],[2,2,1,-1],[1,-1,2,-1],[0,-1,3,4]]

Input: rooms = [[-1]] Output: [[-1]]
Input: rooms = [[2147483647]] Output: [[2147483647]]
Input: rooms = [[0]] Output: [[0]]
*/
class Solution {
    private static final int EMPTY = Integer.MAX_VALUE;
    private static final int GATE = 0;
    private static final List<int[]> DIRECTIONS = Arrays.asList(new int[] { 1, 0 }, new int[] { -1, 0 },
            new int[] { 0, 1 }, new int[] { 0, -1 });

    public void wallsAndGates(int[][] rooms) {
        int m = rooms.length;
        if (m == 0)
            return;
        int n = rooms[0].length;
        Queue<int[]> q = new LinkedList<>();
        for (int row = 0; row < m; row++) {
            for (int col = 0; col < n; col++) {
                if (rooms[row][col] == GATE) {
                    q.add(new int[] { row, col });
                }
            }
        }
        while (!q.isEmpty()) {
            int[] point = q.poll();
            int row = point[0];
            int col = point[1];
            for (int[] direction : DIRECTIONS) {
                int r = row + direction[0];
                int c = col + direction[1];
                if (r < 0 || c < 0 || r >= m || c >= n || rooms[r][c] != EMPTY) {
                    continue;
                }
                rooms[r][c] = rooms[row][col] + 1;
                q.add(new int[] { r, c });
            }
        }
    }
}
// Surrounded Regions
// Shortest Distance from All Buildings
// Robot Room Cleaner

File: Data-Structure/_Algorithm (LeetCode)/Alg-Study-Plan_Algorithm III/Day 7 Breadth-First Search_Depth-First Search/00286. Walls and Gates/286. Walls and Gates.js
// 286. Walls and Gates
/*
You are given an m x n grid rooms initialized with these three possible values.

-1 A wall or an obstacle.
0 A gate.
INF Infinity means an empty room. We use the value 231 - 1 = 2147483647 to represent INF as you may assume that the distance to a gate is less than 2147483647.
Fill each empty room with the distance to its nearest gate. If it is impossible to reach a gate, it should be filled with INF.

Example 1:
Input: rooms = [[2147483647,-1,0,2147483647],[2147483647,2147483647,2147483647,-1],[2147483647,-1,2147483647,-1],[0,-1,2147483647,2147483647]]
Output: [[3,-1,0,1],[2,2,1,-1],[1,-1,2,-1],[0,-1,3,4]]

Example 2:
Input: rooms = [[-1]]
Output: [[-1]]
*/
{
    function dfs(rooms, i, j, dist, isStart, M, N) {
        if (i < 0 || j < 0 || i >= M || j >= N) return;
        if (rooms[i][j] === -1) return;
        if (dist >= rooms[i][j] && !isStart) return;

        rooms[i][j] = dist;

        dfs(rooms, i - 1, j, dist + 1, false, M, N);
        dfs(rooms, i + 1, j, dist + 1, false, M, N);
        dfs(rooms, i, j - 1, dist + 1, false, M, N);
        dfs(rooms, i, j + 1, dist + 1, false, M, N);
    }

    var wallsAndGates = function (rooms) {
        const M = rooms.length;
        const N = rooms[0].length;

        for (var i = 0; i < M; i++) {
            for (var j = 0; j < N; j++) {
                if (rooms[i][j] !== 0) continue;
                dfs(rooms, i, j, 0, true, M, N);
            }
        }
    };
    // Shortest Distance from All Buildings (H)
    // Robot Room Cleaner (H)
}